tests/integration/order/test_creator.py

```diff
@@ -76,6 +77,52 @@ def test_raises_exception_if_duplicate_order_number_passed(self):
                 order_number="1234",
             )
 
+    def test_no_error_with_none_offer_discount(self):
+        """Test that a discount with no offer can be processed without raising an error"""
+        # Setup
+        product = factories.create_product()
+        stockrecord = factories.create_stockrecord(
+            product=product,
+            price=D('10.00'),
+            num_in_stock=1000
+        )
+        self.basket.add_product(product)
+
+        # Calculate shipping and totals
+        shipping_method = Free()
+        shipping_charge = shipping_method.calculate(self.basket)
+        total = calculators.OrderTotalCalculator().calculate(
+            basket=self.basket,
+            shipping_charge=shipping_charge,
+            surcharges=self.surcharges,
+        )
+
+        # Add discount without offer
+        line = self.basket.all_lines()[0]
+        line.discounts.discount(
```

Comment:
```suggestion
        line.discount(
```

Reply:
can't line.discount it crashes, the fonction create_line_discount_model() wait for a list to iterate

Reply:
It does work:
try this:
```suggestion
line.discount(
            discount_value=D('2.00'),
            affected_quantity=1,
            incl_tax=True,
            offer=None
        )
```

Reply:
```
Traceback (most recent call last):
  File "/home/greg/test/django-oscar/tests/integration/order/test_creator.py", line 102, in test_no_error_with_none_offer_discount
    line.discount(
    ~~~~~~~~~~~~~^
        amount=D('2.00'),
        ^^^^^^^^^^^^^^^^^
    ...<2 lines>...
        offer=None
        ^^^^^^^^^^
    )
    ^
TypeError: AbstractLine.discount() got an unexpected keyword argument 'amount'
```

----------------------------------------------------------------------
Ran 1 test in 0.038s

FAILED (errors=1)
Preserving test database for alias 'default'...`

Reply:
```class LineDiscountRegistry(LineOfferConsumer):
    def __init__(self, line):
        super().__init__(line)
        self._discounts = []
        self._discount_excl_tax = None
        self._discount_incl_tax = None

    def discount(self, amount, quantity, incl_tax=True, offer=None):
        self._discounts.append(DiscountApplication(amount, quantity, incl_tax, offer))
        self.consume(quantity, offer=offer)
        if incl_tax:
            self._discount_incl_tax = None
        else:
            self._discount_excl_tax = None```
It's built with that in the business logic: self._discounts.append(DiscountApplication(amount, quantity, incl_tax, offer))

Reply:
> ```
> Traceback (most recent call last):
>   File "/home/greg/test/django-oscar/tests/integration/order/test_creator.py", line 102, in test_no_error_with_none_offer_discount
>     line.discount(
>     ~~~~~~~~~~~~~^
>         amount=D('2.00'),
>         ^^^^^^^^^^^^^^^^^
>     ...<2 lines>...
>         offer=None
>         ^^^^^^^^^^
>     )
>     ^
> TypeError: AbstractLine.discount() got an unexpected keyword argument 'amount'
> 
> 
> ----------------------------------------------------------------------
> Ran 1 test in 0.038s
> 
> FAILED (errors=1)
> Preserving test database for alias 'default'...`
> ```

As the error suggests, amount is not an argument for line's discount method. Please use the code I wrote above (with different parameters i.e, discount_value and affected_quantity), it will work.